File: backend/src/main/java/com/back/ourlog/domain/diary/dto/DiaryResponseDto.java
package com.back.ourlog.domain.diary.dto;

import com.back.ourlog.domain.diary.entity.Diary;

import java.io.Serializable;
import java.time.format.DateTimeFormatter;
import java.util.List;

public record DiaryResponseDto(
        Integer id,
        Integer userId,
        Integer contentId,
        String title,
        String contentText,
        Float rating,
        Boolean isPublic,
        String createdAt,
        String modifiedAt,
        String releasedAt,
        List<String> genres,
        List<String> tags,
        List<String> otts
) implements Serializable {

    public static DiaryResponseDto from(Diary diary) {
        DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");

        return new DiaryResponseDto(
                diary.getId(),
                diary.getUser() != null ? diary.getUser().getId() : null,
                diary.getContent() != null ? diary.getContent().getId() : null,
                diary.getTitle(),
                diary.getContentText(),
                diary.getRating(),
                diary.getIsPublic(),
                diary.getCreatedAt() != null ? diary.getCreatedAt().toString() : null,
                diary.getUpdatedAt() != null ? diary.getUpdatedAt().toString() : null,
                diary.getContent() != null && diary.getContent().getReleasedAt() != null
                        ? diary.getContent().getReleasedAt().format(dateFormatter)
                        : null,
                diary.getDiaryGenres().stream()
                        .map(dg -> dg.getGenre().getName())
                        .toList(),
                diary.getDiaryTags().stream()
                        .map(dt -> dt.getTag().getName())
                        .toList(),
                diary.getDiaryOtts().stream()
                        .map(do_ -> do_.getOtt().getName())
                        .toList()
        );
    }
}

File: backend/src/main/java/com/back/ourlog/domain/diary/dto/DiaryResponseDto.kt
package com.back.ourlog.domain.diary.dto

import com.back.ourlog.domain.diary.entity.Diary
import java.io.Serializable
import java.time.format.DateTimeFormatter

data class DiaryResponseDto(
    val id: Int,
    val userId: Int,
    val contentId: Int,
    val title: String,
    val contentText: String,
    val rating: Float,
    val isPublic: Boolean,
    val createdAt: String,
    val modifiedAt: String,
    val releasedAt: String?,
    val genres: List<String>,
    val tags: List<String>,
    val otts: List<String>
) : Serializable {
    companion object {
        private val DATE_ONLY = DateTimeFormatter.ofPattern("yyyy-MM-dd")

        fun from(diary: Diary): DiaryResponseDto {
            return DiaryResponseDto(
                id = diary.id,
                userId = diary.user.id,
                contentId = diary.content.id,
                title = diary.title,
                contentText = diary.contentText,
                rating = diary.rating,
                isPublic = diary.isPublic,
                createdAt = diary.createdAt.toString(),
                modifiedAt = diary.updatedAt.toString(),
                releasedAt = diary.content.releasedAt?.format(DATE_ONLY),
                genres = diary.diaryGenres.map { it.genre.name },
                tags = diary.diaryTags.map { it.tag.name },
                otts = diary.diaryOtts.map { it.ott.name }
            )
        }
    }
}
